### ISSUE TITLE
test_acceptOutOfFiles mishandles particularly low-fd-availability conditions poorly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4221]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4221|
|Type|defect|
|Created|2010-01-14 22:16:52Z|
|Branch|https://github.com/twisted/twisted/tree/less-confusing-accept-emfile-test-4221|

`test_acceptOutOfFiles` accidentally closes the port it's going to call accept on if the very first non-port socket it tries to create cannot be created because available file descriptors have already been exhausted.  This results in a weird failure later on.  We probably can't easily make the test pass in such a low-fd-availability case, but we can make it fail less confusingly.

<details><summary>Searchable metadata</summary>

```
trac-id__4221 4221
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_less_confusing_accept_emfile_test_4221 branches/less-confusing-accept-emfile-test-4221
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1263507412000000 1263507412000000
changetime__1274739636000000 1274739636000000
version__None None
owner__ 
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28005]) Branching to 'less-confusing-accept-emfile-test-4221'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Yea.  I did it.

With a broken version of pygtk, with this branch and the [#3431](https://github.com/twisted/twisted/issues/3431) branch merged together into trunk, `test_acceptOutOfFiles` fails like this:

```
[ERROR]: twisted.test.test_tcp_internals.PlatformAssumptionsTestCase.test_acceptOutOfFiles

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_tcp_internals.py", line 74, in test_acceptOutOfFiles
    client = self.socket()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_tcp_internals.py", line 57, in socket
    s = socket.socket()
  File "/usr/lib/python2.6/socket.py", line 182, in __init__
    _sock = _realsocket(family, type, proto)
socket.error: [Errno 24] Too many open files
```

This is an improvement over how it fails in the [#3431](https://github.com/twisted/twisted/issues/3431) branch:

```
[FAIL]: twisted.test.test_tcp_internals.PlatformAssumptionsTestCase.test_acceptOutOfFiles

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_tcp_internals.py", line 97, in test_acceptOutOfFiles
    self.assertIn(exc.args[0], (EMFILE, ENOBUFS))
twisted.trial.unittest.FailTest: 9 not in (24, 105)
```

}}}

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess set owner to @exarkun|
|-|-|

On my Mac running 10.6, running ```trial test_tcp_internals.py``` repeatedly will reliably Mac-equivalent-of-blue-screen-of-death my computer. It seems to take between 2 and 5 runs to do this, and it doesn't appear to happen on trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Likewise (after I updated - before that, I couldn't reproduce it).  Filed upstream, along with an example that is even less Twisted-based (ie, doesn't require any bits of Twisted at all).

The Apple bug id is 7596169, but I don't think anyone except Apple employees can do anything with that information.  Let's wait a little while and see what they do, I guess.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

For what it's worth, that Apple bug is now in the state that Apple bugs tend to get in to - I can no longer tell whether or not it has been fixed, and I probably won't receive any more information from Apple about it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [29013]) Skip on OS X

refs [#4221](https://github.com/twisted/twisted/issues/4221)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Letting the OS X kernel bug block this fix any longer is pointless.  Either Apple will fix it someday or they won't.  There's really nothing we can do about it.

Test is marked as skip on OS X now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Okay, in 10.6.3 the OS X bug seems fixed.  So skip removed.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/less-confusing-accept-emfile-test-4221).  Note the Windows failures are from an old gtk bug since fixed in trunk.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @exarkun|
|-|-|

Looks good to me.  Land it, please.

It occurs to me that this test may cause problems in a concurrent testing environment, but we'll have to burn that bridge when we come to it.  Not really relevant to the review, because nothing about this ticket changes that, and I'm sure it's not the only test with that problem.  Let me know if you have any thoughts on it, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [29039]) Merge less-confusing-accept-emfile-test-4221 

Author: exarkun
Reviewer: jesstess, glyph
Fixes: [#4221](https://github.com/twisted/twisted/issues/4221)

Change one of the platform integration tests so that when it fails,
it fails with a more informative error.

Previously `test_acceptOutOfFiles` would fail with an `EBADF` if it
ran out of file descriptors before it managed to allocate any clients.
It did this because it would accidentally close the port it wanted to
accept on.

The test is changed to allocate sockets in a different order now, so
if the same condition is hit, it will fail with a boring `EMFILE` instead.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
